hw/xbox/mcpx_apu.c

```diff
@@ -1478,9 +1528,19 @@ static void se_frame(void *opaque)
     int mixbin;
     int sample;
 
-    timer_mod(d->se.frame_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + 10);
+#if !USE_APU_THREAD
+    timer_mod(d->se.frame_timer, qemu_clock_get_us(QEMU_CLOCK_VIRTUAL) + 600);
+#endif
     MCPX_DPRINTF("mcpx frame ping\n");
 
+    // Check for free space in the output buffer
+    qemu_spin_lock(&d->vp_out_buf_lock);
+    int avail = fifo8_num_free(&d->vp_out_buf);
+    qemu_spin_unlock(&d->vp_out_buf_lock);
+    if (avail < (NUM_SAMPLES_PER_FRAME * 4)) {
+        return;
+    }
+
```

Comment:
We probably want a debug marker here, to ensure that this code path is taken at 1500Hz in the virtual-time.

It requires a comment (or an extension to what's already in line 1523), because the logic implies it will be running at 1500Hz in host-time... which is wrong / leads to problems.

It's not even 1500Hz in host-time, but only average, as there is variable frequency, depending on how large the host audio driver buffer sizes are - so it's possible that it will suddenly consume 20 frames at once, then nothing for a while.
Meanwhile, the virtual-time will still advance, leading to timing issues and various forms of desynchronization (VP / GP, CPU / GP, GP / GP, GP / CPU, GP / EP, EP / EP, EP / ACI - and maybe others I didn't even think of).

The proper solution is to run it at 1500Hz in virtual-time, and [stretch the frames](https://en.wikipedia.org/wiki/Audio_time_stretching_and_pitch_scaling) to 1500Hz in real-time.

Also see discussion in https://github.com/xqemu/xqemu/pull/250#discussion_r303221485

Reply:
Agree, will add a comment about how this throttles to ~1500Hz due to FIFO. At the moment, I'm unsure of how we should handle running at 1500Hz in virtual-time, but it probably doesn't need to be so precise. Once we set up the pipeline to output audio from EP, I think we may not need to worry too much about desync

Reply:
I worry a lot about de-synchronization. There's many things which can go wrong (VP = GP = EP = ACI = Host Audio timing). If any component is too slow (or fast) this will degrade audio quality and potentially affect game logic (GP is in direct contact with CPU and they potentially run in lockstep).

We likely can't guarantee that Host Audio has the same behavior as ACI should have, so we also can't guarantee anything for components feeding ACI (VP / GP / EP). However, at least GP frame timing would potentially affect the CPU = game logic suddenly depends on Host Audio.
This is something we'd want to avoid - Host Audio timing (and effects on logic) is a very prominent issue in most modern emulators.

The easiest solution is to decouple host and guest timing by using time-stretching. That guarantees the same logic for all users, independent from their audio drivers.

It also avoids ugly hacks.

*(Time stretching can be done later of course; just wanted to elaborate as to why I worry about this)*

> Once we set up the pipeline to output audio from EP

Debug audio output from EP or GP doesn't solve the timing issues: you need isochronous 1500Hz, but if EP is hooked up to Host Audio, then you might get something different (Remember VP = GP = EP = ACI = Host Audio timing).

EP outputting audio is also only a temporary hack until we have good ACI.